src/flexmock/api.py

```diff
@@ -311,75 +313,62 @@ def updated(self: Any) -> Any:
                             )
                         )
-                elif expected is not raised:
-                    raise ExceptionClassError('expected "%s", raised "%s"' % (expected, raised))
-            else:
-                raise
+                elif expected_message and expected_message != message:
+                    raise (
+                        ExceptionMessageError(
+                            'expected "%s", raised "%s"' % (expected_message, message)
+                        )
+                    )
+            elif expected is not raised:
+                raise ExceptionClassError('expected "%s", raised "%s"' % (expected, raised))
 
         def match_return_values(expected: Any, received: Any) -> bool:
-            if not isinstance(expected, tuple):
-                expected = (expected,)
-            if not isinstance(received, tuple):
-                received = (received,)
-            if len(received) != len(expected):
-                return False
-            for i, val in enumerate(received):
-                if not _arguments_match(val, expected[i]):
-                    return False
-            return True
+            expected = (expected,) if not isinstance(expected, tuple) else expected
+            received = (received,) if not isinstance(received, tuple) else received
+            return len(received) == len(expected) and all(
+                _arguments_match(r, e) for r, e in zip(received, expected)
+            )
 
         def pass_thru(
             expectation: Expectation, runtime_self: Any, *kargs: Any, **kwargs: Any
         ) -> Any:
-            return_values = None
             try:
-                original = _getattr(expectation, "_original")
-                _mock = _getattr(expectation, "_mock")
-                if inspect.isclass(_mock):
-                    if type(original) in SPECIAL_METHODS:
-                        original = _getattr(expectation, "_original_function")
-                        return_values = original(*kargs, **kwargs)
-                    else:
-                        return_values = original(runtime_self, *kargs, **kwargs)
-                else:
-                    return_values = original(*kargs, **kwargs)
-            except Exception:
+                return_values = (
+                    (
+                        expectation._original_function(*kargs, **kwargs)
+                        if type(expectation._original) in SPECIAL_METHODS
+                        else expectation._original(runtime_self, *kargs, **kwargs)
+                    )
+                    if inspect.isclass(expectation._mock)
+                    else expectation._original(*kargs, **kwargs)
+                )
+            except Exception:  # pylint: disable=broad-except
                 return _handle_exception_matching(expectation)
-            expected_values = _getattr(expectation, "_return_values")
+            expected_values = expectation._return_values
```

Comment:
I wonder what was the need for `_getattr`, I hope we are not breaking anything there … Have you looked up the history ?

Reply:
Note that we wouldn't need the `_getattr` function since it's basically just an alias of `object.__getattribute__`

Reply:
Yeah, we can get rid of the function and call `object.__getattribute__`. However, if we change `Expectation.__getattribute__` like I suggested, then there is no need to call `object.__getattribute__` (expect inside `Expectation.__getattribute__` itself).

Reply:
We could even replace the if-branches with a dict lookup, it would be optimized that way too, or just have those as actual attributes/methods. (I know that it was a choice made before, but I think it takes away from the maintainability of this class anyways.)

Reply:
It seems that they were properties before:
https://github.com/flexmock/flexmock/commit/365c96a33657d720a59bf4f121aa469634b8102a

I think this change was made so that they could be called with or without parenthesis. I think it would be better to have those as actual methods because now code completion doesn't work for them.